README.md
[![ci-cd](https://github.com/johnwslee/extrucal/actions/workflows/ci-cd.yml/badge.svg)](https://github.com/johnwslee/extrucal/actions/workflows/ci-cd.yml)
[![codecov](https://codecov.io/gh/johnwslee/extrucal/branch/main/graph/badge.svg?token=YT37K0ESGF)](https://codecov.io/gh/johnwslee/extrucal)
[![Documentation Status](https://readthedocs.org/projects/extrucal/badge/?version=latest)](https://extrucal.readthedocs.io/en/latest/index.html)
![license
status](https://img.shields.io/github/license/johnwslee/extrucal)

# extrucal

**Author:** John W.S. Lee

`extrucal` provides a class and functions that calculate throughputs and screw RPMs for various types of extrusion processes. Theoretical throughputs can be calculated by using the screw geometry and the processing condition, whereas the throughputs required for extruded products(cable, tube, rod, and sheet) can be calculated by using the product geometry. Based on these calculated throughputs, `extrucal` functions can generate tables and plots that show the processing windows considering extruder size, line speed, and screw RPM.

A large portion of arguments for the functions are given the typical values. Some of the arguments for functions are as follows:
 screw size, channel depth, polymer melt density, screw RPM, screw pitch, flight width, number of flights, line speed, extruder size, etc.

## Installation

`extrucal` can be installed PyPI using the following terminal command:

```bash
$ pip install extrucal
```

## Package Class and Functions

**1. Class and methods in `extrucal.extrusion`**

- `ExtruderScrew()`
  - This is class for an extruder with geometry of the metering section of the extrusion screw

- `ExtruderScrew.throughput()`
  - This method calculates the extrusion throughput (Drag Flow) of the ExtruderScrew object given the melt density of polymer and screw RPM
  
- `ExtruderScrew.throughput_table()`
  - This method generate a table containing the extrusion throughput of the ExtruderScrew object given the melt density and operating conditions
  
- `ExtruderScrew.throughput_plot()`
  - This method generates a plot containing the extrusion throughput of the ExtruderScrew object given the melt density and operating conditions

**2. Functions in `extrucal.extrusion`**

- `throughput_cal()`
  - This function calculates the extrusion throughput (Drag Flow) given the screw size, RPM, the channel depth of metering channel, and screw pitch
  
- `throughput_table()`
  - This function generates a table containing the extrusion throughput with respect to channel depth and screw RPM
  
- `throughput_plot()`
  - This function generates a plot containing the extrusion throughput with respect to channel depth and screw RPM

**3. Functions in `extrucal.cable_extrusion`**

- `cable_cal()`
  - This function calculates the required throughput for cables given the outer diameter, thickness, line speed, and solid polymer density
  
- `cable_table()`
  - This function generate a table containing the required screw RPM with respect to line speed and extruder size
  
- `cable_plot()`
  - This function generate a plot containing the required screw RPM with respect to line speed and extruder size

**4. Functions in `extrucal.tube_extrusion`**

- `tube_cal()`
  - This function calculates the required throughput for tubes given the outer diameter, inner diameter, line speed, and solid polymer density
  
- `tube_table()`
  - This function generate a table containing the required screw RPM with respect to line speed and extruder size
  
- `tube_plot()`
  - This function generate a plot containing the required screw RPM with respect to line speed and extruder size

**5. Functions in `extrucal.rod_extrusion`**

- `rod_cal()`
  - This function calculates the required throughput for rods given the outer diameter, line speed, solid polymer density, and number of die holes
  
- `rod_table()`
  - This function generate a table containing the required screw RPM with respect to line speed and extruder size
  
- `rod_plot()`
  - This function generate a plot containing the required screw RPM with respect to line speed and extruder size

**6. Functions in `extrucal.sheet_extrusion`**

- `sheet_cal()`
  - This function calculates the required throughput for sheets given the width, thickness, line speed, solid polymer density, and number of die holes
  
- `sheet_table()`
  - This function generate a table containing the required screw RPM with respect to line speed and extruder size
  
- `sheet_plot()`
  - This function generate a plot containing the required screw RPM with respect to line speed and extruder size


## Usage

`extrucal` can be used to calculate extrusion throughput and to generate tables and plots of various parameters in extrusion processes

```python
from extrucal.extrusion import ExtruderScrew, throughput_cal, throughput_table, throughput_plot
from extrucal.cable_extrusion import cable_cal, cable_table, cable_plot
from extrucal.tube_extrusion import tube_cal, tube_table, tube_plot
from extrucal.rod_extrusion import rod_cal, rod_table, rod_plot
from extrucal.sheet_extrusion import sheet_cal, sheet_table, sheet_plot
```

## Dependencies

-   Python 3.9 and Python packages:

    -   pandas==1.4.1
    -   numpy==1.22.2
    -   ipykernel==6.9.1
    -   altair-saver==0.5.0
    -   plotly==5.9.0

## Documentation

Documentation `extrucal` can be found at [Read the Docs](https://extrucal.readthedocs.io/en/latest/index.html)

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`extrucal` was created by John W.S. Lee. It is licensed under the terms of the MIT license.

## Credits

`extrucal` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.4.21 (2023-06-14)



## v1.4.20 (2023-06-11)



## v1.4.19 (2023-06-08)



## v1.4.18 (2023-06-08)



## v1.4.17 (2023-06-08)



## v1.4.16 (2023-06-08)



## v1.4.15 (2023-06-08)



## v1.4.14 (2023-06-08)



## v1.4.13 (2023-06-07)



## v1.4.12 (2022-07-21)


## v1.4.11 (2022-02-24)


## v1.4.10 (2022-02-23)


## v1.4.9 (2022-02-23)


## v1.4.8 (2022-02-23)


## v1.4.7 (2022-02-23)


## v1.4.6 (2022-02-23)


## v1.4.5 (2022-02-23)


## v1.4.4 (2022-02-23)


## v1.4.3 (2022-02-23)


## v1.4.2 (2022-02-23)


## v1.4.1 (2022-02-23)


## v1.4.0 (2022-02-23)
### Feature
* Finalize the function sets ([`a07413b`](https://github.com/johnwslee/extrucal/commit/a07413b00b16a1730ac6092daaa674678d4c0901))

### Fix
* Minor modification in test_sheet_extrusion.py ([`2aafd06`](https://github.com/johnwslee/extrucal/commit/2aafd069302eb2cd64e2957d5bcc1eab4d055ab4))

## v1.3.0 (2022-02-23)
### Feature
* Adds additional function ([`b2b29a4`](https://github.com/johnwslee/extrucal/commit/b2b29a4404cf157b9b23071a992aa4b58c1fadd9))

## v1.2.0 (2022-02-23)
### Feature
* Adds functions for tube extrusion ([`55d9f72`](https://github.com/johnwslee/extrucal/commit/55d9f720eec0df0f9901cf9044dd3e76f29bb25b))

## v1.1.0 (2022-02-23)


## v0.2.1 (2022-02-23)


## v0.2.0 (2022-02-22)
### Feature
* Rename extrucal.py and throughput() ([`2f130a5`](https://github.com/johnwslee/extrucal/commit/2f130a5061707ff98c4ce2e9ed2cfbdf2f6de501))

## v0.1.2 (2022-02-22)


## v0.1.1 (2022-02-22)


## v0.1.0 (20/02/2022)

- First release of `extrucal`!

tests/test_rod_extrusion.py
import pytest
import pandas as pd
import numpy as np
import altair as alt
from extrucal.extrusion import throughput_cal
from extrucal.rod_extrusion import rod_cal, rod_table, rod_plot
alt.renderers.enable('html')

def test_input_data():
    """
    Check Errors raised when incorrect inputs are used
    Test cases:
      1. input types
      2. input values
    """
    # Test input types

    with pytest.raises(TypeError):
        rod_cal("5", 10, 1000)
    with pytest.raises(TypeError):
        rod_cal(5, "10", 1000)
    with pytest.raises(TypeError):
        rod_cal(5, 10, "1000")

    with pytest.raises(TypeError):
        rod_table("5", 1000)
    with pytest.raises(TypeError):
        rod_table(5, "1000")
    with pytest.raises(TypeError):
        rod_table(5, 1000, density_ratio="0.85")
    with pytest.raises(TypeError):
        rod_table(5, 1000, min_l_speed="1")
    with pytest.raises(TypeError):
        rod_table(5, 1000, max_l_speed="10")
    with pytest.raises(TypeError):
        rod_table(5, 1000, delta_l_speed="1")
    with pytest.raises(TypeError):
        rod_table(5, 1000, min_size="20")
    with pytest.raises(TypeError):
        rod_table(5, 1000, max_size="100")
    with pytest.raises(TypeError):
        rod_table(5, 1000, delta_size="20")
    with pytest.raises(TypeError):
        rod_table(5, 1000, depth_percent="0.05")

    with pytest.raises(TypeError):
        rod_plot("5", 1000)
    with pytest.raises(TypeError):
        rod_plot(5, "1000")
    with pytest.raises(TypeError):
        rod_plot(5, 1000, density_ratio="0.85")
    with pytest.raises(TypeError):
        rod_plot(5, 1000, min_l_speed="1")
    with pytest.raises(TypeError):
        rod_plot(5, 1000, max_l_speed="10")
    with pytest.raises(TypeError):
        rod_plot(5, 1000, delta_l_speed="1")
    with pytest.raises(TypeError):
        rod_plot(5, 1000, min_size="20")
    with pytest.raises(TypeError):
        rod_plot(5, 1000, max_size="100")
    with pytest.raises(TypeError):
        rod_plot(5, 1000, delta_size="20")
    with pytest.raises(TypeError):
        rod_plot(5, 1000, depth_percent="0.05")

    # Test input values

    with pytest.raises(ValueError):
        rod_cal(5, 10, 299)
    with pytest.raises(ValueError):
        rod_cal(5, 10, 3001)

    with pytest.raises(ValueError):
        rod_table(5, 299)
    with pytest.raises(ValueError):
        rod_table(5, 3001)
    with pytest.raises(ValueError):
        rod_table(5, 1000, density_ratio=1.01)
    with pytest.raises(ValueError):
        rod_table(5, 1000, density_ratio=0.49)
    with pytest.raises(ValueError):
        rod_table(5, 1000, delta_l_speed=10)
    with pytest.raises(ValueError):
        rod_table(5, 1000, delta_size=81)
    with pytest.raises(ValueError):
        rod_table(5, 1000, depth_percent=0.009)
    with pytest.raises(ValueError):
        rod_table(5, 1000, depth_percent=0.31)

    with pytest.raises(ValueError):
        rod_plot(5, 299)
    with pytest.raises(ValueError):
        rod_plot(5, 3001)
    with pytest.raises(ValueError):
        rod_plot(5, 1000, density_ratio=1.01)
    with pytest.raises(ValueError):
        rod_plot(5, 1000, density_ratio=0.49)
    with pytest.raises(ValueError):
        rod_plot(5, 1000, delta_l_speed=10)
    with pytest.raises(ValueError):
        rod_plot(5, 1000, delta_size=81)
    with pytest.raises(ValueError):
        rod_plot(5, 1000, depth_percent=0.009)
    with pytest.raises(ValueError):
        rod_plot(5, 1000, depth_percent=0.31)

def test_output():
    """
    Check if the returned output is correct
    """
    # Test the output of throughput()
    
    expected1 = 117.81
    actual1 = rod_cal(5, 10, 1000, 10)
    assert actual1 == expected1, "Calculated Value is wrong!!!"
    
    expected2 = 188.496
    actual2 = rod_cal(2, 20, 1000, 50)
    assert actual2 == expected2, "Calculated Value is wrong!!!"

    # Test the output of throughput_table()
    
    expected5 = 5
    actual5 = len(rod_table(5, 1000))
    assert actual5 == expected5, "The number of rows doesn't match!!!"
    
    expected6 = 9
    actual6 = len(rod_table(5, 1000, delta_size=10))
    assert actual6 == expected6, "The number of rows doesn't match!!!"
    
    expected8 = 10
    actual8 = len(rod_table(5, 1000).columns)
    assert actual8 == expected8, "The number of columns doesn't match!!!"

    expected9 = 5
    actual9 = len(rod_table(5, 1000, max_l_speed=5).columns)
    assert actual9 == expected9, "The number of columns doesn't match!!!"
    
    # Test the output of throughput_plot()
    
    test_plot = rod_plot(5, 1000)
    assert str(type(test_plot)) == "<class 'altair.vegalite.v4.api.Chart'>"
    assert test_plot.encoding.x.shorthand == 'size', "'size' should be mapped to the x axis"
    assert test_plot.encoding.y.shorthand == 'rpm', "'rpm' should be mapped to the y axis"
    assert test_plot.mark == 'line', "mark should be a line"
    tooltip = "[Tooltip({\n  shorthand: 'size'\n}), Tooltip({\n  shorthand: 'speed'\n}), Tooltip({\n  shorthand: 'rpm'\n})]"
    assert str(rod_plot(5, 1000).encoding.tooltip) == tooltip


src/extrucal/extrusion.py
import altair as alt
import numpy as np
import pandas as pd
import plotly.graph_objects as go


class ExtruderScrew:
    """
    Class for an extruder with geometry of the metering section of the extrusion screw
    """

    def __init__(self, size, depth, pitch=None, w_flight=None, n_flight=1):
        """
        Geometry for the metering section of the screw

        Parameters
        ----------
        size     : int or float
                   Screw size [mm]
        depth    : int or float
                   Channel depth of metering section [mm]
        pitch    : int or float
                   Screw pitch [mm]
                   If None, squared pitch (=1D) is used
        w_flight : int or float
                   Flight width [mm]
                   If None, 10% of screw size is used for flight width
        n_flight : int
                   Number of flight [ea]
                   Default value is 1 (single-flighted)
        """
        self.size = size
        self.depth = depth
        if pitch == None:
            pitch = size
        self.pitch = pitch
        self.w_flight = w_flight
        self.n_flight = n_flight

    def throughput(self, density, rpm):
        """
        Calculates the extrusion throughput (Drag Flow) given the melt density
        of polymer and screw RPM

        Parameters
        ----------
        density  : int or float
                   Melt density of polymeric material [kg/m^3]
        rpm      : int or float
                   Screw RPM
                   Default value is 1 (throughput per unit rpm)
        """
        return throughput_cal(
            size=self.size,
            depth=self.depth,
            density=density,
            rpm=rpm,
            pitch=self.pitch,
            w_flight=self.w_flight,
            n_flight=self.n_flight,
        )

    def throughput_table(self, density, min_rpm=0, max_rpm=100, delta_rpm=10):
        """
        Generate a table containing the extrusion throughput given the melt
        density and operating conditions

        Parameters
        ----------
        density     : int or float
                      Melt density of polymeric material [kg/m^3]
        min_rpm     : int or float
                      Minimum screw RPM for calculation [RPM]
                      Default value is 5 (5RPM)
        max_rpm     : int or float
                      Maximum screw RPM for calculation [RPM]
                      Default value is 50 (50RPM)
        delta_rpm   : int or float
                      Amount of increment in RPM for calculation [RPM]
                      Default value is 5 (5RPM)
        """
        rpms = [rpm for rpm in range(min_rpm, max_rpm, delta_rpm)]
        outputs = [
            round(throughput_cal(
                size=self.size,
                depth=self.depth,
                density=density,
                rpm=rpm,
                pitch=self.pitch,
                w_flight=self.w_flight,
                n_flight=self.n_flight,
            ), 2)
            for rpm in range(min_rpm, max_rpm, delta_rpm)
        ]
        df = pd.DataFrame({"RPM": rpms, "Throughput[kg/hr]": outputs}).set_index("RPM")
        return df

    def throughput_plot(self, density, min_rpm=0, max_rpm=100, delta_rpm=10):
        """
        Generates a plot containing the extrusion throughput given the melt
        density and operating conditions

        Parameters
        ----------
        density     : int or float
                      Melt density of polymeric material [kg/m^3]
        min_rpm     : int or float
                      Minimum screw RPM for calculation [RPM]
                      Default value is 5 (5RPM)
        max_rpm     : int or float
                      Maximum screw RPM for calculation [RPM]
                      Default value is 50 (50RPM)
        delta_rpm   : int or float
                      Amount of increment in RPM for calculation [RPM]
                      Default value is 5 (5RPM)
        """
        rpms = [rpm for rpm in range(min_rpm, max_rpm, delta_rpm)]
        outputs = [
            round(throughput_cal(
                size=self.size,
                depth=self.depth,
                density=density,
                rpm=rpm,
                pitch=self.pitch,
                w_flight=self.w_flight,
                n_flight=self.n_flight,
            ), 2)
            for rpm in range(min_rpm, max_rpm, delta_rpm)
        ]
        df = pd.DataFrame({"RPM": rpms, "Throughput[kg/hr]": outputs})

        fig = go.Figure()
        fig.add_scatter(
            x=df["RPM"],
            y=df["Throughput[kg/hr]"],
            name=f"size={self.size}, depth={self.depth}, pitch={self.pitch}",
            mode="lines+markers",
        )
        fig.update_layout(
            title="Throughput vs RPM",
            xaxis_title="RPM",
            yaxis_title="Throughput[kg/hr]",
            margin=dict(l=20, r=20, t=40, b=20),
            width=400,
            height=300,
        )

        fig.show()


def throughput_cal(size, depth, density, rpm=1, pitch=None, w_flight=None, n_flight=1):
    """
    Calculates the extrusion throughput (Drag Flow) given the screw size, RPM,
    the channel depth of metering channel, and screw pitch

    Parameters
    ----------
    size     : int or float
               Screw size [mm]
    depth    : int or float
               Channel depth of metering section [mm]
    density  : int or float
               Melt density of polymeric material [kg/m^3]
    rpm      : int or float
               Screw RPM
               Default value is 1 (throughput per unit rpm)
    pitch    : int or float
               Screw pitch [mm]
               If None, squared pitch (=1D) is used
    w_flight : int or float
               Flight width [mm]
               If None, 10% of screw size is used for flight width
    n_flight : int
               Number of flight [ea]
               Default value is 1 (single-flighted)

    Returns
    -------
    throughput : float
                 exturion throughput [kg/hr]

    Examples
    --------
    >>> throughput_cal(size=200, depth=10, density=800)
    """

    # Test input type
    if not isinstance(size, int):
        if not isinstance(size, float):
            raise TypeError("'size' should be either integer or float")
    if not isinstance(depth, int):
        if not isinstance(depth, float):
            raise TypeError("'depth' should be either integer or float")
    if not isinstance(density, int):
        if not isinstance(density, float):
            raise TypeError("'density' should be either integer or float")

    # Assign default values
    if pitch == None:
        pitch = size
    if w_flight == None:
        w_flight = size * 0.1

    # Tests the types of default variables
    if not isinstance(rpm, int):
        if not isinstance(rpm, float):
            raise TypeError("'rpm' should be either integer or float")
    if not isinstance(pitch, int):
        if not isinstance(pitch, float):
            raise TypeError("'pitch' should be either integer or float")
    if not isinstance(w_flight, int):
        if not isinstance(w_flight, float):
            raise TypeError("'w_flight' should be either integer or float")
    if not isinstance(n_flight, int):
        raise TypeError("'n_flight' should be integer")

    # Test input value
    if depth < size * 0.01:
        raise ValueError(
            "Channel depth is too shallow(<1% of screw size) to be used for extrusion screw"
        )
    if depth > size * 0.3:
        raise ValueError(
            "Channel depth is too deep(>30% of screw size) to be used for extrusion screw"
        )
    if size < 5:
        raise ValueError("Screw size is too small!!")
    if size > 500:
        raise ValueError("Screw size is too big!!")
    if density < 300:
        raise ValueError("This is not melt density for polymers. Too low!!")
    if density > 3000:
        raise ValueError("This is not melt density for polymers. Too high!!")
    if pitch < size * 0.2:
        raise ValueError("Screw pitch is too small")
    if pitch > size * 2.5:
        raise ValueError("Screw pitch is too big")
    if w_flight < size * 0.01:
        raise ValueError("Flight width is too small")
    if w_flight > size * 0.7:
        raise ValueError("Flight width is too big")
    if n_flight not in [1, 2]:
        raise ValueError(
            "You chose wrong value for n_flight. It should be either 1 or 2"
        )

    # Calculate basic variables
    screw_root_size = size - (depth * 2)
    helix_angle_b = np.arctan(pitch / (np.pi * size))
    helix_angle_c = np.arctan(pitch / (np.pi * screw_root_size))
    channel_width_b = ((pitch / n_flight) * np.cos(helix_angle_b)) - w_flight
    channel_width_c = ((pitch / n_flight) * np.cos(helix_angle_c)) - w_flight
    avg_channel_width = (channel_width_b + channel_width_c) / 2

    # Generates table for shape factor (Drag)
    f_d_dict = dict()
    idx = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25]
    for i in idx:
        f_d_dict[i] = (1 / (i**3)) * np.tanh(
            (i * np.pi * depth) / (2 * avg_channel_width)
        )
    f_d_df = pd.Series(f_d_dict)

    # Calculation of basic variables continues
    shape_factor_drag = ((16 * avg_channel_width) / (np.pi**3 * depth)) * f_d_df.sum()
    rotation_per_sec = rpm / 60
    barrel_rot_speed = (np.pi * rotation_per_sec * size * np.cos(helix_angle_b)) / 1000
    throughput_per_sec = (
        n_flight
        * density
        * barrel_rot_speed
        * (avg_channel_width / 1000)
        * (depth / 1000)
        * shape_factor_drag
    ) / 2
    throughput_per_hr = throughput_per_sec * 60 * 60

    return throughput_per_hr


def throughput_table(
    size,
    density,
    pitch=None,
    w_flight=None,
    n_flight=1,
    min_depth=None,
    max_depth=None,
    delta_depth=None,
    min_rpm=5,
    max_rpm=50,
    delta_rpm=5,
):
    """
    Generate a table containing the extrusion throughput with respect to
    channel depth and screw RPM

    Parameters
    ----------
    size        : int or float
                  Screw size [mm]
    density     : int or float
                  Melt density of polymeric material [kg/m^3]
    pitch       : int or float
                  Screw pitch [mm]
                  If None, squared pitch (=1D) is used
    w_flight    : int or float
                  Flight width [mm]
                  If None, 10% of screw size is used for flight width
    n_flight    : int
                  Number of flight [ea]
                  Default value is 1 (single-flighted)
    min_depth   : int or float
                  Minimum depth for calculation [mm]
                  If None, 2% of screw size is used for minimum depth
    max_depth   : int or float
                  Maximum depth for calculation [mm]
                  If None, 9% of screw size is used for maximum depth
    delta_depth : int or float
                  Amount of increment in depth for calculation [mm]
                  If None, 1% of screw size is used for depth increment
    min_rpm     : int or float
                  Minimum screw RPM for calculation [RPM]
                  Default value is 5 (5RPM)
    max_rpm     : int or float
                  Maximum screw RPM for calculation [RPM]
                  Default value is 50 (50RPM)
    delta_rpm   : int or float
                  Amount of increment in RPM for calculation [RPM]
                  Default value is 5 (5RPM)

    Returns
    -------
    table : pandas.DataFrame
            dataframe containing the throughput as a function of channel depth and screw RPM

    Examples
    --------
    >>> output_table(size=200, density=800)
    """

    # Test input type
    if not isinstance(size, int):
        if not isinstance(size, float):
            raise TypeError("'size' should be either integer or float")
    if not isinstance(density, int):
        if not isinstance(density, float):
            raise TypeError("'density' should be either integer or float")

    # Assign default values
    if pitch == None:
        pitch = size
    if w_flight == None:
        w_flight = size * 0.1
    if min_depth == None:
        min_depth = size * 0.02
    if max_depth == None:
        max_depth = size * 0.09
    if delta_depth == None:
        delta_depth = size * 0.01

    # Tests the types of default variables
    if not isinstance(pitch, int):
        if not isinstance(pitch, float):
            raise TypeError("'pitch' should be either integer or float")
    if not isinstance(w_flight, int):
        if not isinstance(w_flight, float):
            raise TypeError("'w_flight' should be either integer or float")
    if not isinstance(n_flight, int):
        raise TypeError("'n_flight' should be integer")
    if not isinstance(min_depth, int):
        if not isinstance(min_depth, float):
            raise TypeError("'min_depth' should be either integer or float")
    if not isinstance(max_depth, int):
        if not isinstance(max_depth, float):
            raise TypeError("'max_depth' should be either integer or float")
    if not isinstance(delta_depth, int):
        if not isinstance(delta_depth, float):
            raise TypeError("'delta_depth' should be either integer or float")
    if not isinstance(min_rpm, int):
        if not isinstance(min_rpm, float):
            raise TypeError("'min_rpm' should be either integer or float")
    if not isinstance(max_rpm, int):
        if not isinstance(max_rpm, float):
            raise TypeError("'max_rpm' should be either integer or float")
    if not isinstance(delta_rpm, int):
        if not isinstance(delta_rpm, float):
            raise TypeError("'delta_rpm' should be either integer or float")

        # Test input value
    if min_depth < size * 0.01:
        raise ValueError(
            "Channel depth is too shallow(<1% of screw size) to be used for extrusion screw"
        )
    if max_depth > size * 0.3:
        raise ValueError(
            "Channel depth is too deep(>30% of screw size) to be used for extrusion screw"
        )
    if delta_depth > max_depth - min_depth:
        raise ValueError(
            "'delta_depth' can not be greater than 'max_depth - min_depth'"
        )
    if size < 5:
        raise ValueError("Screw size is too small!!")
    if size > 500:
        raise ValueError("Screw size is too big!!")
    if density < 300:
        raise ValueError("This is not melt density for polymers. Too low!!")
    if density > 3000:
        raise ValueError("This is not melt density for polymers. Too high!!")
    if pitch < size * 0.2:
        raise ValueError("Screw pitch is too small")
    if pitch > size * 2.5:
        raise ValueError("Screw pitch is too big")
    if w_flight < size * 0.01:
        raise ValueError("Flight width is too small")
    if w_flight > size * 0.7:
        raise ValueError("Flight width is too big")
    if n_flight not in [1, 2]:
        raise ValueError(
            "You chose wrong value for n_flight. It should be either 1 or 2"
        )

    # Generates table
    table = dict()
    throughput_list = list()
    depth = [round(i, 2) for i in np.arange(min_depth, max_depth + 0.1, delta_depth)]
    rpm = [j for j in np.arange(min_rpm, max_rpm + 0.1, delta_rpm)]
    rpm_title = [f"rpm={k}" for k in np.arange(min_rpm, max_rpm + 0.1, delta_rpm)]
    for d in depth:
        for r in rpm:
            throughput_list.append(
                round(throughput_cal(size, d, density, r, pitch, w_flight, n_flight), 2)
            )
        table[f"depth={d}"] = throughput_list
        throughput_list = []
    table_df = pd.DataFrame(table, index=rpm_title)

    return table_df


def throughput_plot(
    size,
    density,
    pitch=None,
    w_flight=None,
    n_flight=1,
    min_depth=None,
    max_depth=None,
    delta_depth=None,
    min_rpm=0,
    max_rpm=50,
    delta_rpm=1,
):
    """
    Generates a plot containing the extrusion throughput with respect to
    channel depth and screw RPM

    Parameters
    ----------
    size        : int or float
                  Screw size [mm]
    density     : int or float
                  Melt density of polymeric material [kg/m^3]
    pitch       : int or float
                  Screw pitch [mm]
                  If None, squared pitch (=1D) is used
    w_flight    : int or float
                  Flight width [mm]
                  If None, 10% of screw size is used for flight width
    n_flight    : int
                  Number of flight [ea]
                  Default value is 1 (single-flighted)
    min_depth   : int or float
                  Minimum depth for calculation [mm]
                  If None, 2% of screw size is used for minimum depth
    max_depth   : int or float
                  Maximum depth for calculation [mm]
                  If None, 9% of screw size is used for maximum depth
    delta_depth : int or float
                  Amount of increment in depth for calculation [mm]
                  If None, 1% of screw size is used for depth increment
    min_rpm     : int or float
                  Minimum screw RPM for calculation [RPM]
                  Default value is 0 (0RPM)
    max_rpm     : int or float
                  Maximum screw RPM for calculation [RPM]
                  Default value is 50 (50RPM)
    delta_rpm   : int or float
                  Amount of increment in RPM for calculation [RPM]
                  Default value is 1 (1RPM)

    Returns
    -------
    plot : altair.Chart object
           a line chart showing the throughput as a function of channel depth and screw RPM

    Examples
    --------
    >>> output_plot(size=200, density=800)
    """

    # Test input type
    if not isinstance(size, int):
        if not isinstance(size, float):
            raise TypeError("'size' should be either integer or float")
    if not isinstance(density, int):
        if not isinstance(density, float):
            raise TypeError("'density' should be either integer or float")

    # Assign default values
    if pitch == None:
        pitch = size
    if w_flight == None:
        w_flight = size * 0.1
    if min_depth == None:
        min_depth = size * 0.02
    if max_depth == None:
        max_depth = size * 0.09
    if delta_depth == None:
        delta_depth = size * 0.01

    # Tests the types of default variables
    if not isinstance(pitch, int):
        if not isinstance(pitch, float):
            raise TypeError("'pitch' should be either integer or float")
    if not isinstance(w_flight, int):
        if not isinstance(w_flight, float):
            raise TypeError("'w_flight' should be either integer or float")
    if not isinstance(n_flight, int):
        raise TypeError("'n_flight' should be integer")
    if not isinstance(min_depth, int):
        if not isinstance(min_depth, float):
            raise TypeError("'min_depth' should be either integer or float")
    if not isinstance(max_depth, int):
        if not isinstance(max_depth, float):
            raise TypeError("'max_depth' should be either integer or float")
    if not isinstance(delta_depth, int):
        if not isinstance(delta_depth, float):
            raise TypeError("'delta_depth' should be either integer or float")
    if not isinstance(min_rpm, int):
        if not isinstance(min_rpm, float):
            raise TypeError("'min_rpm' should be either integer or float")
    if not isinstance(max_rpm, int):
        if not isinstance(max_rpm, float):
            raise TypeError("'max_rpm' should be either integer or float")
    if not isinstance(delta_rpm, int):
        if not isinstance(delta_rpm, float):
            raise TypeError("'delta_rpm' should be either integer or float")

    # Test input value
    if min_depth < size * 0.01:
        raise ValueError(
            "Channel depth is too shallow(<1% of screw size) to be used for extrusion screw"
        )
    if max_depth > size * 0.3:
        raise ValueError(
            "Channel depth is too deep(>30% of screw size) to be used for extrusion screw"
        )
    if delta_depth > max_depth - min_depth:
        raise ValueError(
            "'delta_depth' can not be greater than 'max_depth - min_depth'"
        )
    if size < 5:
        raise ValueError("Screw size is too small!!")
    if size > 500:
        raise ValueError("Screw size is too big!!")
    if density < 300:
        raise ValueError("This is not melt density for polymers. Too low!!")
    if density > 3000:
        raise ValueError("This is not melt density for polymers. Too high!!")
    if pitch < size * 0.2:
        raise ValueError("Screw pitch is too small")
    if pitch > size * 2.5:
        raise ValueError("Screw pitch is too big")
    if w_flight < size * 0.01:
        raise ValueError("Flight width is too small")
    if w_flight > size * 0.7:
        raise ValueError("Flight width is too big")
    if n_flight not in [1, 2]:
        raise ValueError(
            "You chose wrong value for n_flight. It should be either 1 or 2"
        )

    # Generate table for plot
    table = dict()
    throughput_list = list()
    depth = [round(i, 2) for i in np.arange(min_depth, max_depth + 0.1, delta_depth)]
    rpm = [j for j in np.arange(min_rpm, max_rpm + 0.1, delta_rpm)]
    for d in depth:
        for r in rpm:
            throughput_list.append(
                round(throughput_cal(size, d, density, r, pitch, w_flight, n_flight), 2)
            )
        table[d] = throughput_list
        throughput_list = []
    table_df = pd.DataFrame(table, index=rpm)
    table_for_plot = table_df.reset_index()
    table_for_plot = table_for_plot.rename(columns={"index": "RPM"})
    table_for_plot = table_for_plot.melt(
        id_vars="RPM", var_name="depth", value_name="throughput"
    )
    table_for_plot["depth"] = table_for_plot["depth"].astype("category")

    # Generate plot
    plot = (
        alt.Chart(table_for_plot, title="Throughput vs Screw RPM & Channel Depth")
        .mark_circle()
        .encode(
            alt.X("RPM", title="Screw RPM", scale=alt.Scale(domain=(0, max_rpm))),
            alt.Y("throughput", title="Throughput [kg/hr]"),
            alt.Color(
                "depth",
                title="Channel depth [mm]",
                sort=alt.EncodingSortField("throughput", op="mean", order="descending"),
            ),
            tooltip=["RPM", "depth", "throughput"],
        )
        .configure_axis(labelFontSize=14, titleFontSize=16)
        .configure_legend(labelFontSize=16, titleFontSize=14)
        .configure_title(fontSize=18)
    )
    return plot